crate_universe/src/select.rs

```diff
@@ -0,0 +1,594 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::Debug;
+
+use serde::{de::DeserializeOwned, Deserialize, Deserializer, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+pub struct Select<T>
+where
+    T: Selectable,
+{
+    common: T::CommonType,
+    selects: BTreeMap<String, T::SelectsType>,
+}
+
+pub trait Selectable
+where
+    Self: SelectableValue,
+{
+    type ItemType: SelectableValue;
+    type CommonType: SelectableValue + Default;
+    type SelectsType: SelectableValue;
+
+    fn is_empty(this: &Select<Self>) -> bool;
+    fn insert(this: &mut Select<Self>, value: Self::ItemType, configuration: Option<String>);
+
+    fn items(this: &Select<Self>) -> Vec<(Option<String>, Self::ItemType)>;
+    fn values(this: &Select<Self>) -> Vec<Self::ItemType>;
+
+    fn merge(lhs: Select<Self>, rhs: Select<Self>) -> Select<Self>;
+}
+
+// Replace with `trait_alias` once stabilized.
+// https://github.com/rust-lang/rust/issues/41517
+pub trait SelectableValue
+where
+    Self: Debug + Clone + PartialEq + Eq + PartialOrd + Ord + Serialize + DeserializeOwned,
```

Comment:
Hmm partial complication here, the implementation of remapping configs involves building a `BTreeMap<T, BTreeSet<old config>>` which means that `T` always requires `Ord`.  Any preference on how you'd like to handle that?  Look at the `new()` functions in `select_list.rs`, etc.

Reply:
I don't think I see `T` being used as a key in any `BTreeMap`s there? I see:
```rust
let mut remapped: BTreeMap<String, Vec<(T, String)>>;
let mut unmapped: Vec<(T, String)>;

#[derive(Debug, PartialEq, Eq)]
pub struct SelectList<T>
where
    T: SelectableValue,
{
    common: Vec<T>,
    selects: BTreeMap<String, Vec<WithOriginalConfigurations<T>>>,
    unmapped: Vec<WithOriginalConfigurations<T>>,
}
```

But all of those look like they use `T` either as a value, or as a `Vec` entry, not as a a `BTreeMap` key?

Reply:
Sorry, bad example.  `select_list.rs` is the only one that uses a `Vec` for unmapped.  The others are all `BTreeMap`.

My understanding being that for all the others, if multiple configs are unmapped, we merge identical values to produce a single value with multiple "original configurations".  For the `SelectList` case, there's no deduplication/merging/whatever across configs, so they all just show up separately.  It could deduplicate on the entire `Vec` instead of each individual item.  I could also just not do unmapped in a `BTreeMap` and just search for the value in a `Vec`, though that potentially has ordering stability issues.

Thoughts?  Sorry I responded to this so late today.

Reply:
Can we add the trait bounds in each of the other types?

`SelectList` _not_ having this bound is really the thing I'm hoping for, I understand it is probably necessary everywhere else :)